pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
+		} else {
+			notIngestedCounts.count(ts)
+		}
+	}
+
+	// Truncate the toIngest slice to the index up to which we wrote TimeSeries data into it,
+	// all the TimeSeries objects beyond the write index must be returned to the pool.
+	for _, ts := range tsSlice[tsSliceWriteIdx:] {
+		f.pools.putTs(ts.TimeSeries)
+	}
+	tsSlice = tsSlice[:tsSliceWriteIdx]
 
-type recoverableError struct {
-	error
+	return notIngestedCounts, tsSlice, tsByTargets
 }
 
-func (r *request) Send(ctx context.Context) <-chan error {
-	errCh := make(chan error, 1)
+func (f *forwarder) growTimeseriesSlice(ts []mimirpb.PreallocTimeseries) []mimirpb.PreallocTimeseries {
+	newPos := len(ts)
 
-	// Early return if there's no data to send.
-	if len(r.tsByEndpoint) == 0 {
-		close(errCh)
-		return errCh
+	if cap(ts) > len(ts) {
+		ts = ts[:newPos+1]
+	} else {
+		ts = append(ts, mimirpb.PreallocTimeseries{})
 	}
 
-	returnErr := func(err error) {
-		if !r.propagateErrors {
-			return
-		}
-		errCh <- err
-	}
+	ts[newPos].TimeSeries = f.pools.getTs()
 
-	var wg sync.WaitGroup
-	wg.Add(len(r.tsByEndpoint))
+	return ts
+}
 
-	// Can't use concurrency.ForEachJob because we don't want to cancel the other jobs if one errors.
-	errorsByEndpoint := make(map[string]error, len(r.tsByEndpoint))
-	var errorsMtx sync.Mutex
-	for endpoint, ts := range r.tsByEndpoint {
-		go func(endpoint string, ts []mimirpb.PreallocTimeseries) {
-			defer wg.Done()
+// worker is a worker go routine which performs the forwarding requests that it receives through a channel.
+func (f *forwarder) worker() {
+	defer f.workerWg.Done()
 
-			errorsMtx.Lock()
-			defer errorsMtx.Unlock()
-			errorsByEndpoint[endpoint] = r.sendToEndpoint(ctx, endpoint, ts)
-		}(endpoint, *ts)
+	for req := range f.reqCh {
+		req.do()
 	}
+}
 
-	go func() {
-		defer r.cleanup()
-		defer close(errCh)
+type request struct {
+	pools  *pools
+	client *http.Client
+	log    log.Logger
 
-		wg.Wait()
+	ctx             context.Context
+	timeout         time.Duration
+	propagateErrors bool
+	errCh           chan error
+	requestWg       *sync.WaitGroup
 
-		var nonRecoverable error
-		// No need to get errorsMtx because we already waited for all routines which might modify it to end.
-		for endpoint, err := range errorsByEndpoint {
-			if err == nil {
-				continue
-			}
+	endpoint string
+	ts       tsWithSampleCount
 
-			if errors.As(err, &recoverableError{}) {
-				// If there is at least one recoverable error we want to return the recoverable error.
-				returnErr(httpgrpc.Errorf(http.StatusInternalServerError, "endpoint %s: %s", endpoint, err.Error()))
-				return
-			}
+	requests  prometheus.Counter
+	errors    *prometheus.CounterVec
+	samples   prometheus.Counter
+	exemplars prometheus.Counter
+	latency   prometheus.Histogram
+}
 
-			nonRecoverable = httpgrpc.Errorf(http.StatusBadRequest, "endpoint %s: %s", endpoint, err.Error())
-		}
+// newRequest launches a new forwarding request and sends it to a worker via a channel.
```

Comment:
`newRequest` makes me think this is going to create a new request object and return it. Maybe something like `submitForwardingJob()` ?

Reply:
good point, fixed here: 83357caff44df54ad2a46680eecfe262ed735a3f